Wrappers/Python/cil/framework/framework.py

```diff
@@ -2481,7 +2481,61 @@ def maximum(self, x2, *args, **kwargs):
     def minimum(self,x2, out=None, *args, **kwargs):
         return self.pixel_wise_binary(numpy.minimum, x2=x2, out=out, *args, **kwargs)
 
+
+    def sapyb(self, a, y, b, out=None, dtype=numpy.float32, num_threads=NUM_THREADS):
+        '''performs a*self + b * y. It is inline safe
+        
+        Parameters
+        ----------
+        a : multiplier for self, can be a number or a numpy array or a DataContainer
+        y : DataContainer 
+        b : multiplier for y, can be a number or a numpy array or a DataContainer
+        out : return DataContainer, if None a new DataContainer is returned, default None. 
+            out can be self or y.
+        dtype : forces the output to the type, default numpy float32
+        num_threads : number of threads to use during the calculation, using the CIL C library
+        
+        It will try to use the CIL C library and default to numpy operations, in case the C library does
+        not handle the types.
+        
+        Example:
+        -------
+
+        a = 2
+        b = 3
+        ig = ImageGeometry(10,11)
+        x = ig.allocate(1)
+        y = ig.allocate(2)
+        out = x.sapyb(a,y,b)
+        '''
+        use_c_lib = True
+        ret_out = False
+        if self.dtype in numpy.sctypes['complex']:
+            use_c_lib = False
+        
+        if out is None:
+            out = self * 0.
+            ret_out = True
+
```

Comment:
The c implementation of `axpby` can work with `x` and `y` being arrays of floats or doubles, and `a` and `b` being either scalars or arrays  of floats or doubles respectively.

The dtype used is controlled by the dtype of `out`, and a `TypeError` is raised if this isn't float32 or float64
https://github.com/TomographicImaging/CIL/blob/0e78d20fa31f7ec90e18107ba5e4ef5a575d00d0/Wrappers/Python/cil/framework/framework.py#L2611

So I'd suggest creating `out`, then if dtype is float32 or float64 call `_axpyb` for the c code and otherwise reverting to python. This means we only pass to C the datatypes we are guaranteed to accept, whereas as it is it checks for complex only. 

We could also try to catch the runtime error of the C code and default to python if this happens?
https://github.com/TomographicImaging/CIL/blob/0e78d20fa31f7ec90e18107ba5e4ef5a575d00d0/Wrappers/Python/cil/framework/framework.py#L2637

The RunTime error  could be caused by anything :) but most likely memory access as we don't check on the array length passed - in which case numpy will fail too I assume.

I know you won't like changing requested behaviour with only a warning and not a crash though!

Reply:
It could still crash in the fallback method, no?